DEV-2302: Add "publish" button to global header in page editor and fire "publish page" modal

GuiMend opened the pull request:
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)
N/A

#### What's this PR do?
Moves the Publish/Unpublish logic from Edit sidebar to global header so it's easier to find and publish pages.

#### Why are we doing this? How does it help us?
This change makes it clearer for the user to find, publish & unpublish Donation Pages.

#### How should this be manually tested? Please include detailed step-by-step instructions.
- Login with a user that has at least 1 donation page
- Enter edit mode for donation page
- If page is already Published, global header should show "Published" button next to "Grab Link"
- If page is unpublished, global header should show "Publish" button alone
	- Clicking through Publish flow should publish page, and update global header buttons state.
- Check that all the designs match figma (both in dev url and in storybook)
- Unpublish functionality disabled for now
![image](https://user-images.githubusercontent.com/9381325/189939245-77bfe46f-4aad-4811-a975-513424f25295.png)

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?
Yes

#### Have automated unit tests been added? If not, why?
yes

#### How should this change be communicated to end users?
Publish/Unpublish flow is now much easier to use, edit and see.

#### Are there any smells or added technical debt to note?
No

#### Has this been documented? If so, where?
No

#### What are the relevant tickets? Add a link to any relevant ones.
https://news-revenue-hub.atlassian.net/browse/DEV-2302
https://news-revenue-hub.atlassian.net/browse/DEV-2323

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?
No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:
https://news-revenue-hub.atlassian.net/browse/DEV-2298
https://news-revenue-hub.atlassian.net/browse/DEV-2304
https://news-revenue-hub.atlassian.net/browse/DEV-2299

nrh-cklimas reviewed (changes requested):
Overall looks good! Just questions here and there.

GuiMend commented on spa/cypress/integration/02 - donation-page-edit.spec.js:
```diff
@@ -410,19 +410,12 @@ describe('Donation page edit', () => {
+      cy.getByTestId('preview-page-button').click();
+      cy.get(`button[aria-label="Published page ${livePage.name}"]`).click();
+      cy.contains(expectedFormat);
+      cy.get('body').click(0, 0);
```
Yes

GuiMend commented on spa/src/components/common/Button/PublishButton/PublishButton.js:
```diff
@@ -0,0 +1,152 @@
+  );
+
+  const handleUnpublish = () => {
+    patchPage({ published_date: addYears(new Date(), 10) });
```
That's a good question, will bring it on slack. From what I've seen the current implementation is that we don't have an "unpublish", we can only schedule a publish date in the future. Will double check just to be certain.

nrh-cklimas replied:
I think looking at it, the intent is to remove unpublish functionality for now? If so I think we should hide the button--if I understand the code as it is, it'll display but do nothing.

GuiMend replied:
Removed it 👍

GuiMend commented on spa/src/components/common/Button/PublishButton/PublishButton.js:
```diff
@@ -0,0 +1,152 @@
+import { Flex, Button, Popover, LiveText, UnpublishButton, Text } from './PublishButton.styled';
+import PublishModal from '../PublishModal';
+
+const PublishButton = ({ page, setPage, className, alert, requestPatchPage }) => {
```
That's correct, it will blink for published pages while data loads. I haven't seen a loading state for the header, this is probably a question for Rachel or Mariclare

GuiMend replied:
Removed the publish button while page is loading 👍

GuiMend commented on spa/src/components/common/Button/PublishModal/PublishModal.js:
```diff
@@ -0,0 +1,118 @@
+            </Grid>
+            <Grid item xs={2} />
+            <Grid item xs={5}>
+              {/* TODO: add input error messages */}
```
TODO in a follow up ticket for improvements

gurjittoptal reviewed (approved):
approving. ( as no longer in this one )

GuiMend commented on spa/src/components/common/Button/PublishButton/PublishButton.styled.js:
```diff
@@ -0,0 +1,106 @@
+export const UnpublishButton = styled(MuiButton)`
+  && {
+    color: white;
+    width: 123px;
```
Not sure where you are seeing the widht not being applied, but it's there because when we click "unpublish", a loading state will show and if I don't fix the widht it will shrink (and I want it to stay the same width)
![image](https://user-images.githubusercontent.com/9381325/190432573-ddf4ade0-5b59-4d81-857f-d81145b39c0b.png)
![image](https://user-images.githubusercontent.com/9381325/190432909-aa266ca3-4602-40c8-af15-3e9cbfceec79.png)

nrh-cklimas replied:
I see. I was looking at the button before you click on it. Something we should follow up on when we start to build out the style guide.

GuiMend commented on spa/src/components/common/Button/PublishModal/PublishModal.styled.js:
```diff
@@ -0,0 +1,146 @@
+  font-size: ${(props) => props.theme.fontSizesUpdated.xs};
+  color: ${(props) => props.theme.colors.muiGrey[600]};
+  font-weight: 400;
+  margin: 8px 0 0 0 !important;
```
I tried messing around but couldn't override the style without important 😞

nrh-cklimas replied:
Weird. I can show you what I see locally on a screenshare. Margin definitely seems to require !important but padding doesn't for me.

GuiMend replied:
I couldn't make the element have the same design as Figma by only changing the padding. The defaut `margin: 20px 0;` need to be removed, that's why I added the important

GuiMend commented on spa/src/components/common/Button/PublishModal/PublishModal.js:
```diff
@@ -0,0 +1,118 @@
+  };
+
+  return (
+    <Modal open={open} onClose={onClose} aria-labelledby={`Publish page ${page?.name}`}>
```
That's right, updated it to `aria-label` + strengthen unit test

GuiMend commented on spa/src/components/common/Button/PublishButton/PublishButton.js:
```diff
@@ -0,0 +1,152 @@
+            variant="contained"
+            onClick={isPublished ? handleOpenPopover : handleOpen}
+            active={showPopover ? 'true' : ''}
+            aria-label={`${isPublished ? 'Published' : 'Publish'} page ${page?.name}`}
```
Done

GuiMend commented:
@nrh-cklimas , addressed all your comments on [fb050a5](https://github.com/newsrevenuehub/rev-engine/pull/665/commits/fb050a51e1bf83fc07ec127d7651c00b1950c953)

nrh-cklimas commented on spa/src/components/common/Button/PublishModal/PublishModal.js:
```diff
@@ -0,0 +1,118 @@
+              <Label>Page Name</Label>
+            </Grid>
+            <Grid item xs={4}>
+              <Input value={page?.revenue_program?.slug} start="true" onChange={() => {}} />
```
If it's not editable, can we put a `readonly` prop on the input instead of ignoring input in JS?

I see the same Jest error--weird that only Jest flags this but I guess jsdom is stricter than real life. I think the issue here is we're applying a non-standard prop, `start`, on a vanilla `<input>`. Something we can address when we standardize these lower level components.

GuiMend replied:
Added `readOnly`.

yes, that is the problem, non-standard prop.

nrh-cklimas commented on spa/src/components/content/pages/AddPageModal.js:
```diff
@@ -40,7 +40,7 @@ function AddPageModal({ isOpen, closeModal }) {
   const [name, setName] = useState('');
   const [slug, setSlug] = useState('');
```
I think we can lessen some complexity here now that we know the user will never set the slug manually. Here's what I'm thinking...

- Remove `slug` state since it's always derived from `name`.
- Remove `handleNameBlur`.
- Move the logic of setting the slug to `handleSave`.

I might be missing something here though that would prevent us from doing this, though.

GuiMend replied:
Correct, just don't think we need to do that because a follow up ticket will remove the modal altogether

nrh-cklimas commented:
Looks good, just needs a merge/rebase from main.
